File: js_port/fs_js/src/funcs/list/SortListFunction.js
const { BaseFunction, CallType } = require('../../core/functionBase');
const { ArrayFsList } = require('../../model/FsList');
const helpers = require('../helpers');
const { FsError } = helpers;
const { FSDataType } = require('../../core/fstypes');

class SortListFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'Sort';
    this.callType = CallType.Dual;
  }

  get maxParameters() {
    return 2;
  }

  evaluate(provider, parameters) {
    const error = helpers.expectParamCount(this.symbol, parameters, this.maxParameters);
    if (error) {
      return error;
    }

    const list = helpers.ensureList(parameters.getParameter(provider, 0));
    if (!list) {
      return helpers.makeError(
        FsError.ERROR_TYPE_MISMATCH,
        `${this.symbol} function: The first parameter should be ${this.parName(0)}`
      );
    }

    const fn = helpers.ensureFunction(parameters.getParameter(provider, 1));
    if (!fn) {
      return helpers.makeError(
        FsError.ERROR_TYPE_MISMATCH,
        `${this.symbol} function: The second parameter should be ${this.parName(1)}`
      );
    }

    const copy = [];
    for (let i = 0; i < list.length; i += 1) {
      copy.push(list.get(i));
    }

    copy.sort((left, right) => {
      const args = new helpers.ArrayParameterList([left, right]);
      const result = fn.evaluate(provider, args);
      const typedResult = helpers.ensureTyped(result);
      if (helpers.typeOf(typedResult) !== FSDataType.Integer) {
        throw new Error(`${this.symbol} function: sorting function must return an integer`);
      }
      return helpers.valueOf(typedResult);
    });

    return helpers.makeValue(FSDataType.List, new ArrayFsList(copy));
  }

  parName(index) {
    return index === 0 ? 'List' : index === 1 ? 'Sorting Function' : '';
  }
}

module.exports = {
  SortListFunction
};

File: funcscript/Functions/List/SortListFunction.cs
using FuncScript.Core;
using FuncScript.Model;
using System.Collections.Generic;
using System.Linq;

namespace FuncScript.Functions.List
{
    public class SortListFunction : IFsFunction
    {
        public int MaxParsCount => 2;

        public CallType CallType => CallType.Dual;

        public string Symbol => "Sort";

        public int Precedence => 0;

        public object Evaluate(IFsDataProvider parent, IParameterList pars)
        {
            if (pars.Count != this.MaxParsCount)
                throw new Error.TypeMismatchError($"{this.Symbol} function: Invalid parameter count. Expected {this.MaxParsCount}, but got {pars.Count}");

            var par0 = pars.GetParameter(parent,0);
            var par1 = pars.GetParameter(parent,1);

            return EvaluateInternal(parent, par0, par1);
        }

        private object EvaluateInternal(IFsDataProvider parent, object par0, object par1)
        {
            if (par0 == null)
                return null;

            if (!(par0 is FsList))
                throw new Error.TypeMismatchError($"{this.Symbol} function: The first parameter should be {this.ParName(0)}");

            if (!(par1 is IFsFunction))
                throw new Error.TypeMismatchError($"{this.Symbol} function: The second parameter should be {this.ParName(1)}");

            var func = (IFsFunction)par1;
            var lst = (FsList)par0;
            var res = new List<object>(lst);

            res.Sort((x, y) =>
            {
                var sortParamList = new ArrayParameterList(new object[] { x, y });
                var result = func.Evaluate(parent, sortParamList);

                if (!(result is int))
                    throw new Error.EvaluationTimeException($"{this.Symbol} function: The sorting function must return an integer");

                return (int)result;
            });

            return new ArrayFsList(res);
        }


        public string ParName(int index)
        {
            return index switch
            {
                0 => "List",
                1 => "Sorting Function",
                _ => ""
            };
        }
    }
}
